parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt

```diff
@@ -0,0 +1,158 @@
+            FakeIntervalCalculator()
+        )
+    }
+
+    @Test
+    fun `when starting manager, then record the correct event after interval millis`() {
+        // when
+        sut.start()
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val timestamp = now - THREAD_SLEEPING_THRESHOLD
+
+        // then
+        assertThat(tracker.events[0]).isCorrectEvent(
+            withTotalTime = {
+                // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS
+                isCloseTo(DEFAULT_INTERVAL_MILLIS, withinPercentage(10))
+            },
+            withTimestamp = {
+                // Ideally: timestamp should be equal to System.currentTimeMillis() at the time of recording the event
+                isCloseTo(
+                    timestamp,
+                    within(100L)
+                )
+            }
+        )
+    }
+
+    @Test
+    fun `when starting manager, then schedule task each interval period`() {
+        sut.start()
+
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val firstTimestamp = now - THREAD_SLEEPING_THRESHOLD
+
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val secondTimestamp = now - 2 * THREAD_SLEEPING_THRESHOLD
+
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val thirdTimestamp = now - 3 * THREAD_SLEEPING_THRESHOLD
+
+        sleep(THREAD_SLEEPING_THRESHOLD)
+
+        val firstEvent = tracker.events[0]
+        assertThat(firstEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(firstTimestamp, within(20L)) }
+        )
+        val secondEvent = tracker.events[1]
+        assertThat(secondEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 2
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 2, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(secondTimestamp, within(20L)) }
+        )
+        val thirdEvent = tracker.events[2]
+        assertThat(thirdEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 3
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 3, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(thirdTimestamp, within(20L)) }
+        )
+    }
+
+    private fun MapAssert<String, Any>.isCorrectEvent(
+        withTotalTime: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+        withTimestamp: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+    ): MapAssert<String, Any> {
+        return containsEntry("action", "heartbeat")
```

Comment:
Minor (🔍): Consider moving the `.hasEntrySatisfying("data")` check above the `.containsEntry("inc", 0L)` check to make it easier to follow the same logical order found within the `EngagementManager.doEnqueue(...)` method, that is, without having to jump up and down to verify test correctness.

```kotlin
        return containsEntry("action", "heartbeat")
            // Incremental will be always 0 because the interval is lower than 1s
            .hasEntrySatisfying("data") { data ->
                data as Map<String, Any>
                assertThat(data).hasEntrySatisfying("ts") { timestamp ->
                    timestamp as Long
                    assertThat(timestamp).withTimestamp()
                }.containsAllEntriesOf(testData.minus("ts"))
            }.containsEntry("inc", 0L)
            .hasEntrySatisfying("tt") { totalTime ->
                totalTime as Long
                assertThat(totalTime).withTotalTime()
            }
```